### PR TITLE
SQL-668: bsontype can be an array or a single String.

### PR DESCRIPTION
**Pull request by nbagnard**
This PR handles decoding schema where bsontype is either a Set<String> or String.
It also implements a logic for transforming sets of types in bsontype in a list of anyOf and flattening anyOfs.

The assumptions I made for this are:
- `{
    "bsonType": "bool",
    "additionalProperties": false
}`
is not equals to 
`{
    "bsonType": "bool",
    "additionalProperties": true
}`
so the same type can be twice in the anyOf list if the other fields values are different.
- If bsonType is a set of types, there will be no duplicates. For example, no `bsontype: [string, object, bool, object]`

### CODE REVIEW

File: src/main/java/com/mongodb/jdbc/JsonSchema.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.bson.BsonValue;
+import org.bson.codecs.pojo.annotations.BsonIgnore;
+
+import static java.lang.String.format;
+
+// Simple POJO for deserializing jsonschema.
+public class JsonSchema {
+    private BsonValue polymorphicBsonType;
+    @BsonIgnore
+    public Set<String> bsonType;
```
Comment (rychipman) [inline src/main/java/com/mongodb/jdbc/JsonSchema.java:22]:
```suggestion
    public BsonValue bsonType;
```
I like the idea of having this `JsonSchema` class (for use as a deserialization target) as something separate from `MongoJsonSchema` (for using everywhere else in the API). I'd suggest that we go even further towards just making this class just a "dumb" deserialization target:
- keep the raw BsonValue for the `bsonType` field
- put whatever array/string polymophism-handling logic we need in the function that converts this to a `MongoJsonSchema`

**Review by rychipman — CHANGES_REQUESTED** [review]
only one relatively minor request from me

**Comment by pmeredit** [comment]
> This PR handles decoding schema where bsontype is either a Set or String. It also implements a logic for transforming sets of types in bsontype in a list of anyOf and flattening anyOfs.
> 
> The assumptions I made for this are:
> 
> * `{ "bsonType": "bool", "additionalProperties": false }`
>   is not equals to
>   `{ "bsonType": "bool", "additionalProperties": true }`
>   so the same type can be twice in the anyOf list if the other fields values are different.

It's still correct to have the same type twice, so I think this is a good simplifying assumption.

> * If bsonType is a set of types, there will be no duplicates. For example, no `bsontype: [string, object, bool, object]`

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 2]
```diff
@@ -4,22 +4,196 @@
...
+            for (JsonSchema baseAnyOf : baseSchema.anyOf) {
+                result.anyOf.add(toSimplifiedMongoJsonSchema(baseAnyOf));
+            }
+        }
+        result.required = baseSchema.required;
+        result.items = toSimplifiedMongoJsonSchema(baseSchema.items);
+        result.additionalProperties = baseSchema.additionalProperties;
+
+        if (baseSchema.bsonType!= null) {
+            Set<String> bsonTypes = polymorphicBsonTypeToStringSet(baseSchema.bsonType);
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:72]: I love this method name.

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 3]
```diff
@@ -4,22 +4,196 @@
...
+            //  If there are many types in the set and it can not be reduced to one type after eliminating any Null
+            //  type in the list, the types will be inserted in the list  of anyOf to be handled as polymorphic type.
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:74]:
```suggestion
            //  type in the list, the types will be inserted in the list of anyOf to be handled as polymorphic type.
```

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 4]
```diff
@@ -4,22 +4,196 @@
...
+                }
+            }
+
+            // We'll need to add the types from bsontype set to the anyOf list.
+            // If the list is null, create a new one.
+            if (null == result.anyOf) {
+                result.anyOf = new HashSet<MongoJsonSchema>();
+            }
+
+            // Push down each bsontype into its own anyOf schema
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:102]: Could you add an example here in the comment to clarify what this mean?

**Review by pmeredit — CHANGES_REQUESTED** [review]
I still need to go over the latter half of this. The first half looks good, just a few small requests.

**Comment by pmeredit** [comment]
As a small comment, could we avoid removing new lines from the end of all the files? It messes up the PR. If that is something we want to do, we should do it in a separate PR. But I assume this was some IDE setting :)

File: src/main/java/com/mongodb/jdbc/JsonSchema.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    public JsonSchema items;
+    public boolean additionalProperties;
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof MongoJsonSchema)) {
+            return false;
+        }
+        MongoJsonSchema other = (MongoJsonSchema) obj;
+        return Objects.equals(bsonType, other.bsonType)
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/JsonSchema.java:24]: Is `equals` defined for `BsonValue`?

Comment (pmeredit) [reply src/main/java/com/mongodb/jdbc/JsonSchema.java:24]: Yes, it is!

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 6]
```diff
@@ -4,22 +4,216 @@
...
+             * "y": {
+             *     "bsonType": ["string", "int"]
+             *  }
+             *  will become
+             *  "y": {
+             *      "anyOf": [
+             *          {"bsonType": "string"},
+             *          {"bsonType": "int"}
+             *      ]
+             *  }
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:114]: perfect!

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 7]
```diff
@@ -4,22 +4,216 @@
...
+                }
+            }
+
+            // We'll need to add the types from bsontype set to the anyOf list.
+            // If the list is null, create a new one.
+            if (null == result.anyOf) {
+                result.anyOf = new HashSet<MongoJsonSchema>();
+            }
+
+            /** Push down each bsontype into its own anyOf schema.
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:103]:
```suggestion
            /** Push down each bsontype from a bsontype array into a separate bsonType in an anyOf schema.
```

I think my problem with this comment before is "its own anyOf" which seems like we would have one `anyOf` per `bsonType`.

File: src/main/java/com/mongodb/jdbc/MongoJsonSchema.java [hunk 8]
```diff
@@ -4,22 +4,216 @@
...
+        if (null == ioSchema) {
+            return null;
+        }
+
+        MongoJsonSchema result = ioSchema;
+        if (result.anyOf != null && !result.anyOf.isEmpty()) {
+            result.anyOf =
+                    result.anyOf
+                            .stream()
+                            .flatMap(
```
Comment (pmeredit) [inline src/main/java/com/mongodb/jdbc/MongoJsonSchema.java:191]: <3 `flatMap`. Combinators just make code so much easier to follow once you learn them.

**Review by pmeredit — APPROVED** [review]
LGTM! Feel free to apply the comment suggestion or not.

**Review by rychipman — APPROVED** [review]
LGTM

**Pull request merged**